concent_api/core/tests/test_payments_report.py

```diff
@@ -0,0 +1,789 @@
+import base64
+import mock
+
+from django.test import override_settings
+from golem_messages import message
+from golem_messages.cryptography import ECCx
+
+from common.constants import ConcentUseCase
+from common.helpers import deserialize_message
+from core.constants import MOCK_TRANSACTION_HASH
+from core.message_handlers import store_subtask
+from core.models import Client
+from core.models import DepositAccount
+from core.models import DepositClaim
+from core.models import PaymentInfo
+from core.models import PendingResponse
+from core.models import Subtask
+from core.payments.report import create_report
+from core.payments.report import get_subtasks_with_distinct_clients_pairs
+from core.payments.report import get_errors_and_warnings
+from core.payments.report import MatchingError
+from core.payments.report import MatchingWarning
+from core.payments.report import Payments
+from core.payments.report import PendingResponses
+from core.tests.utils import ConcentIntegrationTestCase
+from core.transfer_operations import store_pending_message
+from core.utils import get_current_utc_timestamp
+from core.utils import hex_to_bytes_convert
+from core.utils import parse_timestamp_to_utc_datetime
+
+
+@override_settings(
+    ADDITIONAL_VERIFICATION_COST=1,
+)
+class PaymentsReportTest(ConcentIntegrationTestCase):
+
+    def setUp(self):
+        super().setUp()
+        self.providers = []
+        self.requestors = []
+        self.subtasks = []
+        self.get_list_of_payments_mocks = []
+        self.get_covered_additional_verification_costs_mocks = []
+        self.closure_time = get_current_utc_timestamp()
+        self.payment_ts = parse_timestamp_to_utc_datetime(self.closure_time)
+        self.default_payment_amount = 10
+        self.number_of_clients_per_type = 10
+        self.number_of_subtasks = 30
+        # The value below is determined by approach of mixing requestors and providers in _prepare_subtasks
+        self.number_of_pairs = self.number_of_clients_per_type * 2
+
+        self._prepare_providers()
+        self._prepare_requestors()
+        self._prepare_subtasks()
+        self._prepare_related_payments_mocks()
+        self._prepare_related_pending_responses()
+
+        assert self.default_payment_amount > 0
+        assert self.number_of_clients_per_type > 1
+        assert self.number_of_subtasks > 0
+
+    def test_create_report(self):
```

Comment:
this test is very complex and contains too much logic. Quite frankly, I cannot understand what is being done here...